tao_compiler/mlir/disc/transforms/disc_hlo_legalize_to_lhlo.cc

```diff
@@ -218,8 +220,96 @@ struct HloToLhloCustomCallOpV2Converter
       resultTypes.push_back(
           MemRefType::get(ty.getShape(), ty.getElementType()));
     }
+
     rewriter.replaceOpWithNewOp<lmhlo_disc::CustomCallV2Op>(
         hloOp, resultTypes, adaptor.getOperands(), hloOp->getAttrs());
+
+    return success();
+  }
+};
+
+struct CustomCallOpConverter : public BaseOpConversion<mhlo::CustomCallOp> {
+ public:
+  using BaseOpConversion<mhlo::CustomCallOp>::BaseOpConversion;
+
+  LogicalResult matchAndRewrite(
+      mhlo::CustomCallOp hloOp, OpAdaptor adaptor,
+      ConversionPatternRewriter& rewriter) const override {
+    Operation* op = hloOp.getOperation();
+    auto operands = adaptor.getOperands();
+
+    SmallVector<Type> resultTypes;
+
+    std::string input_placements, output_placements;
+    for (int i = 0; i < operands.size(); i++) {
+      input_placements += "d,";
+    }
+
+    hloOp->setAttr("call_target_name",
+                   rewriter.getStringAttr(hloOp.getCallTargetName()));
+    hloOp->setAttr("device", rewriter.getStringAttr("x"));
+    hloOp->setAttr("input_layouts", rewriter.getStringAttr("*"));
+    hloOp->setAttr("output_layouts", rewriter.getStringAttr("*"));
+    hloOp->setAttr("expected_input_layouts", rewriter.getStringAttr("*"));
+    hloOp->setAttr("expected_output_layouts", rewriter.getStringAttr("*"));
+
+    SmallVector<NamedAttribute> newAttrs;
+    if (hloOp.getBackendConfig().has_value()) {
+      newAttrs.push_back(
+          NamedAttribute(rewriter.getStringAttr("backend_config"),
+                         hloOp.getBackendConfig().value()));
+    }
+    auto newCustomAttrs = DictionaryAttr::get(hloOp->getContext(), newAttrs);
+    hloOp->setAttr("custom_attrs", newCustomAttrs);
+
+    if (hloOp->getNumResults() == 1 &&
```

Comment:
how do we deal with two results operand of `tuple<>, Value` ? Maybe we can return a failure message  If we don't want to deal with this situation.

Reply:
> how do we deal with two results operand of `tuple<>, Value` ? Maybe we can return a failure message If we don't want to deal with this situation.

It will fall back to normal processing pipeline and emit error when trying to collect result types.

Reply:
Fixed